### PR TITLE
Refactor: 🍰 Change the query so that we only look on the ``contributions`` table.

### LINKED ISSUE #2202: 🔧 [Refactor] Union Query in `util/creation.ts`
## 🔧 Refactor ticket
As we have a confirmed contribution for all creation transactions, we do not need a union any more. Write the query as typorm query.
Include pending means that we also read the contributions that are not confirmed yet. Without pending contributions confirmed at must be null.
Please write some extensive tests for this query. Create lots of contributions for the last three months, and see that errors are thrown if expected. Confirm some contributions and test, that everything behaves as expected.

### LINKED ISSUE #2298: 🐛 [Bug] Cannot confirm contribution
## 🐛 Bugreport

Trying to confirm an users transaction results in this error:
image

The user has enough GDD to be created available
image

User has alot of small contributions confirm
image

### LINKED ISSUE #2299: 🐛 [Bug] Automatic link contributions can exceed maximum amount of monthly GDD
## 🐛 Bugreport

Automatic link contributions can exceed maximum amount of monthly GDD

image
https://media.discordapp.net/attachments/933717057172832276/1033350691797864548/Bildschirmfoto_2022-10-22_um_14.01.41.png?width=1920&height=902

Search for `joy` on production and see account of `Bernd von Gradido`. The screenshots are from that account, but on production the data is completely different - to be clarified

### PR DESCRIPTION
**Pull request by Elweyn**
## 🍰 Pullrequest

In the process of analysing errors we got an error on the open creations due to an error in the SQL Query.
Fix SQL Query to get datas from contributions table only.

### Issues

- fixes #2202
- fixes #2299
- fixes #2298 

### Todo

- [X] Refactor SQL query to typeorm query
- [x] Write tests

### CODE REVIEW

File: backend/src/graphql/resolver/TransactionLinkResolver.ts [hunk 1]
```diff
@@ -258,7 +258,7 @@ export class TransactionLinkResolver {
           }
         }
 
-        const creations = await getUserCreation(user.id, false)
+        const creations = await getUserCreation(user.id)
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/TransactionLinkResolver.ts:261]: why is the field exported in singular when the function definition itself is plural (trailing s). This seems inconsistent

File: backend/src/graphql/resolver/TransactionLinkResolver.test.ts [hunk 2]
```diff
@@ -160,6 +171,56 @@ describe('TransactionLinkResolver', () => {
           ],
         })
       })
+
+      describe('after one day', () => {
+        beforeAll(async () => {
+          jest.useFakeTimers()
+          /* eslint-disable-next-line @typescript-eslint/no-empty-function */
+          setTimeout(() => {}, 1000 * 60 * 60 * 24)
+          jest.runAllTimers()
```
Comment (clauspeterhuebner) [inline backend/src/graphql/resolver/TransactionLinkResolver.test.ts:180]: What in detail do these three lines of code before?

File: backend/src/graphql/resolver/util/creations.ts [hunk 3]
```diff
@@ -50,37 +49,36 @@ export const getUserCreations = async (
...
+    .createQueryBuilder(Contribution, 'c')
+    .select('month(contribution_date)', 'month')
+    .addSelect('user_id', 'userId')
+    .addSelect('sum(amount)', 'sum')
+    .where(`user_id in (${ids.toString()})`)
+    .andWhere(`contribution_date >= ${dateFilter}`)
+    .andWhere('deleted_at IS NULL')
+    .andWhere('denied_at IS NULL')
+    .groupBy('month')
+    .addGroupBy('userId')
```
Comment (clauspeterhuebner) [inline backend/src/graphql/resolver/util/creations.ts:62]: is the order by date DESC not necessary any more?

**Review by clauspeterhuebner — APPROVED** [review]
Some open Question, but if it works like expected then i approve it.

File: backend/src/graphql/resolver/TransactionLinkResolver.test.ts (hunk 2 above)
Comment (Mogge) [reply backend/src/graphql/resolver/TransactionLinkResolver.test.ts:180]:
```javascript
// say the test suite that we want to use a faked clock
jest.useFakeTimers()
// a void function that should run after 24h
setTimeout(() => {}, 1000 * 60 * 60 * 24)
// let all timers, e.g. the void function above, run 
jest.runAllTimers()
// Now the clock is 24h later
```

File: backend/src/graphql/resolver/TransactionLinkResolver.test.ts [hunk 4]
```diff
@@ -160,6 +171,56 @@ describe('TransactionLinkResolver', () => {
...
+            }),
+          ).resolves.toMatchObject({
+            errors: [
+              new GraphQLError(
+                'Creation from contribution link was not successful. Error: Contribution link already redeemed today',
+              ),
+            ],
+          })
+        })
+      })
```
[the remark this answers is not in the document]
Comment (Mogge) [reply backend/src/graphql/resolver/TransactionLinkResolver.test.ts:223]: The link is validated against server time. I have no idea how to redeem a link in the past

File: backend/src/graphql/resolver/util/creations.ts (hunk 3 above)
Comment (Mogge) [reply backend/src/graphql/resolver/util/creations.ts:62]: You are right. Good catch. Thank you!

**Review by ogerly — APPROVED** [review]

**Pull request merged**